port/unix/omrsysinfo.c

```diff
@@ -6495,140 +6490,159 @@ readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFl
-						portLibrary,
-						OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_READ_FAILED,
-						"fgets failed to read %s file stream with errno=%d",
-						OMR_PROC_PID_ONE_SCHED_FILE,
-						osErrCode);
+						portLibrary, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_FOPEN_FAILED,
+						"fopen failed to open %s file with errno=%d", OMR_PROC_PID_ONE_SCHED_FILE, osErrCode);
 				goto _end;
 			}
-		} else {
-			/* Check if the initialization process is either "init " or "systemd ".
-			 * The space after systemd and init allows us to exactly check for those
-			 * strings and filter any strings which either begin with systemd or init.
-			 *
-			 * The first line of /proc/1/sched in a non-containerized environment:
-			 * - systemd (1, #threads: 1)
-			 * - init (1, #threads: 1)
-			 *
-			 * The first line of /proc/1/sched in a containerized environment:
-			 * - bash (1, #threads: 1)
-			 * - sh (1, #threads: 1)
-			 */
+
+			if (NULL == fgets(buffer, sizeof(buffer), schedFile)) {
+				if (0 != ferror(schedFile)) {
+					int32_t osErrCode = errno;
+					Trc_PRT_isRunningInContainer_fgets_failed(OMR_PROC_PID_ONE_SCHED_FILE, osErrCode);
+					rc = portLibrary->error_set_last_error_with_message_format(
+							portLibrary, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_READ_FAILED,
+							"fgets failed to read %s file stream with errno=%d",
+							OMR_PROC_PID_ONE_SCHED_FILE, osErrCode);
+					goto _end;
+				}
+			} else {
+				/* Check if the initialization process is either "init " or "systemd ".
+				 * The space after systemd and init allows us to exactly check for those
+				 * strings and filter any strings which either begin with systemd or init.
+				 *
+				 * The first line of /proc/1/sched in a non-containerized environment:
+				 * - systemd (1, #threads: 1)
+				 * - init (1, #threads: 1)
+				 *
+				 * The first line of /proc/1/sched in a containerized environment:
+				 * - bash (1, #threads: 1)
+				 * - sh (1, #threads: 1)
+				 */
 #define STARTS_WITH(string, prefix) (0 == strncmp(string, prefix, sizeof(prefix) - 1))
-			if (!STARTS_WITH(buffer, "init ") && !STARTS_WITH(buffer, "systemd ")) {
-				*inContainer = TRUE;
-			}
+				if (!STARTS_WITH(buffer, "init ") && !STARTS_WITH(buffer, "systemd ")) {
+					inContainer = TRUE;
+				}
 #undef STARTS_WITH
+			}
 		}
-	}
 
-	Trc_PRT_isRunningInContainer_container_detected((uintptr_t)*inContainer);
 _end:
-	if (NULL != cgroupFile) {
-		fclose(cgroupFile);
+		if (NULL != cgroupFile) {
+			fclose(cgroupFile);
+		}
+
+		if (0 != rc) {
+			setAtomic(&PPG_isRunningInContainer, OMRPORT_RUNNING_IN_CONTAINER_ERROR);
+			/* Print a warning message. */
+			portLibrary->nls_printf(
+					portLibrary, J9NLS_WARNING, J9NLS_PORT_RUNNING_IN_CONTAINER_FAILURE,
+					omrerror_last_error_message(portLibrary));
+		} else if (inContainer) {
+			setAtomic(&PPG_isRunningInContainer, OMRPORT_RUNNING_IN_CONTAINER_TRUE);
+		} else {
+			setAtomic(&PPG_isRunningInContainer, OMRPORT_RUNNING_IN_CONTAINER_FALSE);
```

Comment:
`setAtomic` is not appropriate here. Two (or more) threads might be trying to concurrently determine whether we're in a container. If they reach different conclusions, more care is needed to ensure a reasonable value is stored.

Reply:
Replaced `setAtomic` with a `monitor`. Renamed an existing monitor for this purpose.